kof-operator/internal/controller/configmap_controller.go

```diff
@@ -0,0 +1,576 @@
+
+	// Get the input `ConfigMaps`.
+	alertConfigMaps, err := r.getConfigMaps(ctx, releaseNamespace, KofAlertRulesClusterNameLabel)
+	if err != nil {
+		return err
+	}
+	recordConfigMaps, err := r.getConfigMaps(ctx, releaseNamespace, KofRecordRulesClusterNameLabel)
+	if err != nil {
+		return err
+	}
+
+	// Get the output `ConfigMaps`.
+	// TODO: Revisit namespaces after multi-tenancy is implemented.
+	// These `ConfigMaps` are owned by `ClusterDeployments` that may be in a different namespace.
+	vmRuleConfigMaps, err := r.getConfigMaps(ctx, "", KofRecordVMRulesClusterNameLabel)
+	if err != nil {
+		return err
+	}
+
+	// Merge alert and record `PrometheusRules` into the nested maps.
+	err = r.mergePrometheusRules(
+		ctx,
+		releaseNamespace,
+		releaseName,
+		clusterGroupAlertRules,
+		clusterGroupRecordRules,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Merge alert and record `ConfigMaps` into the nested maps.
+	err = mergeAlertConfigMaps(ctx, alertConfigMaps, clusterGroupAlertRules)
+	if err != nil {
+		return err
+	}
+	err = mergeRecordConfigMaps(ctx, recordConfigMaps, clusterGroupRecordRules)
+	if err != nil {
+		return err
+	}
+
+	// Update `kof-mothership-promxy-rules` ConfigMap with the files from the nested map.
+	alertFiles, err := getAlertFiles(ctx, clusterGroupAlertRules)
+	if err != nil {
+		return err
+	}
+	err = r.updateConfigMap(ctx, nil, releaseNamespace, releaseName+"-promxy-rules", alertFiles)
+	if err != nil {
+		return err
+	}
+
+	// Update each `kof-record-vmrules-$regional_cluster_name` ConfigMap from the nested map.
+	for _, vmRuleConfigMap := range vmRuleConfigMaps {
+		err := r.updateRecordVMRulesConfigMap(ctx, clusterGroupRecordRules, &vmRuleConfigMap)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get `ConfigMaps` with the given `label` and optional `namespace`.
+// Default `ConfigMaps` are moved to the beginning of the list,
+// as we want to merge them first.
+func (r *ConfigMapReconciler) getConfigMaps(
+	ctx context.Context,
+	namespace string,
+	label string,
+) ([]corev1.ConfigMap, error) {
+	log := log.FromContext(ctx)
+
+	options := []client.ListOption{client.HasLabels{label}}
+	if namespace != "" {
+		options = append(options, client.InNamespace(namespace))
+	}
+
+	configMapList := &corev1.ConfigMapList{}
+	if err := r.List(ctx, configMapList, options...); err != nil {
+		log.Error(
+			err, "failed to list ConfigMaps",
+			"label", label,
+			"namespace", namespace,
+		)
+		return nil, err
+	}
+	configMaps := configMapList.Items
+
+	defaultConfigMaps := make([]corev1.ConfigMap, 0, len(configMaps))
+	clusterConfigMaps := make([]corev1.ConfigMap, 0, len(configMaps))
+	for _, configMap := range configMaps {
+		if configMap.Labels[label] == DefaultClusterName {
+			defaultConfigMaps = append(defaultConfigMaps, configMap)
+		} else {
+			clusterConfigMaps = append(clusterConfigMaps, configMap)
+		}
+	}
+	return append(defaultConfigMaps, clusterConfigMaps...), nil
```

Comment:
What is the point of splitting defaultConfigMaps and ClusterConfigMaps if eventually a combined array is returned?